File: rust/morth/src/parse.rs
use crate::{op::Op, token::Token, Error, Result, Value};

pub(crate) fn parse_token_as_op(tok: Token) -> Result<Op> {
    match tok.text.as_str() {
        "+" => Ok(Op::plus(tok.location)),
        "-" => Ok(Op::minus(tok.location)),
        "." => Ok(Op::dump(tok.location)),
        _ => {
            let value = tok
                .text
                .parse::<Value>()
                .map_err(|_| Error::UnknownToken(tok.text))?;
            Ok(Op::push(value, tok.location))
        }
    }
}

File: c++/source/morth/parse.cpp
#include "morth/parse.hpp"

#include <charconv>
#include <fmt/format.h>
#include <stdexcept>

Op ParseTokenAsOp(Token tok) {
    if (tok.text == "+") {
        return Plus(tok);
    }
    if (tok.text == "-") {
        return Minus(tok);
    }
    if (tok.text == "dump") {
        return Dump(tok);
    }

    Int64 value;
    auto result = std::from_chars(tok.text.data(), tok.text.data() + tok.text.size(), value);
    if (result.ec != std::errc{}) {
        fmt::print(stderr, "{}:{}:{}: Invalid token '{}'\n", tok.location.filePath,
                   tok.location.row, tok.location.column, tok.text);
        throw std::runtime_error{"Invalid token"};
    }

    return Push(value, tok);
}
